src/Scripting/Core/ScriptMetadataResolver.cs

```diff
@@ -64,8 +66,30 @@ public ScriptMetadataResolver WithBaseDirectory(string? baseDirectory)
 
         public override bool ResolveMissingAssemblies => _resolver.ResolveMissingAssemblies;
 
-        public override PortableExecutableReference? ResolveMissingAssembly(MetadataReference definition, AssemblyIdentity referenceIdentity)
-            => _resolver.ResolveMissingAssembly(definition, referenceIdentity);
+        public override PortableExecutableReference? ResolveMissingAssembly(MetadataReference definition,
+            AssemblyIdentity referenceIdentity)
+        {
+            bool lockTaken = false;
+            try
+            {
+                Monitor.TryEnter(_lockObject, 100, ref lockTaken);
```

Comment:
It isn't acceptable to limit concurrency like this. The logic is also very problematic and can cause infinite recursion in case of slow IO.

The issue should be fixed in mono itself, or potentially workarounded in MonoGlobalAssemblyCache.